src/centroiders.cpp

```diff
@@ -1,95 +1,199 @@
-}
-
-void CentroidComparison::Print() {
-    printf("Extra stars: %d\nMissing stars: %d\nMean error: %f\n",
-           numExtraStars, numMissingStars, meanError);
-}
-
-}
+#include "centroiders.hpp"
+
+#include <stdio.h>
+#include <stdlib.h>
+#include <math.h>
+
+#include <vector>
+#include <iostream>
+
+#include <unordered_map>
+
+#include <unordered_set>
+
+namespace lost {
+
+// DUMMY
+
+std::vector<Star> DummyCentroidAlgorithm::Go(unsigned char *image, int imageWidth, int imageHeight) const {
+    std::vector<Star> result;
+
+    for (int i = 0; i < numStars; i++) {
+        result.push_back(Star(rand() % imageWidth, rand() % imageHeight, 10.0));
+    }
+
+    return result;
+}
+
+//recursive helper here
+void cogHelper(float &yCoordMagSum,  
+    float &xCoordMagSum, 
+    int &magSum, 
+    int &xMin, 
+    int &xMax, 
+    int &yMin, 
+    int &yMax, 
+    int &cutoff, 
+    std::unordered_set<int> &checkedIndeces, int i, unsigned char *image, int imageWidth, int imageHeight) {
+    if (i >= 0 && i < imageWidth * imageHeight && image[i] >= cutoff && checkedIndeces.count(i) == 0) {
+        checkedIndeces.insert(i);
+        if (i % imageWidth > xMax) {
+            xMax = i % imageWidth;
+        } else if (i % imageWidth < xMin) {
+            xMin = i % imageWidth;
+        }
+        if (i / imageWidth > yMax) {
+            yMax = i / imageWidth;
+        } else if (i / imageWidth < yMin) {
+            yMin = i / imageWidth;
+        }
+        magSum += image[i];
+        xCoordMagSum += ((i % imageWidth) + 1) * image[i];
+        yCoordMagSum += ((i / imageWidth) + 1) * image[i];
+        if((i + 1) % imageWidth != 0) {
+            cogHelper(yCoordMagSum, xCoordMagSum, magSum, xMin, xMax, yMin, yMax, cutoff, checkedIndeces, i + 1, image, imageWidth, imageHeight);
+        }
+        if ((i - 1) % imageWidth != (imageWidth - 1)) {
+            cogHelper(yCoordMagSum, xCoordMagSum, magSum, xMin, xMax, yMin, yMax, cutoff, checkedIndeces, i - 1, image, imageWidth, imageHeight);
+        }
+        cogHelper(yCoordMagSum, xCoordMagSum, magSum, xMin, xMax, yMin, yMax, cutoff, checkedIndeces, i + imageWidth, image, imageWidth, imageHeight);
+        cogHelper(yCoordMagSum, xCoordMagSum, magSum, xMin, xMax, yMin, yMax, cutoff, checkedIndeces, i - imageWidth, image, imageWidth, imageHeight);
+    }
+}
+
+std::vector<Star> CenterOfGravityAlgorithm::Go(unsigned char *image, int imageWidth, int imageHeight) const {
+    float yCoordMagSum = 0; 
+    float xCoordMagSum = 0;
+    int magSum = 0;
+    int xMin;
+    int xMax;
+    int yMin;
+    int yMax;
+    int cutoff;
+    std::unordered_set<int> checkedIndeces;
+    
+    std::vector<Star> result;
+    //loop through entire array, find sum of magnitudes
+    int totalMag = 0;
+    for (int i = 0; i < imageHeight * imageWidth; i++) {
+        totalMag += image[i];
+    }
+    // cutoff might need a new equation
+    cutoff = (((totalMag/(imageHeight * imageWidth)) + 1) * 15) / 10;

```

Comment:
Split cutoff calculation into another function so we can use it from other centroiding algorithms?

Reply:
Done - Defined cutoffFunction as a global function that can be used in other algorithms